File: atcoder.jp/PyPy3 (7.3.0)/abc312/abc312_d.py
S = list(input())
N = len(S)

MOD = 998244353

dp = [[0] * (N + 1) for _ in range(N + 1)]
dp[0][0] = 1

for idx in range(N):
    for cnt in range(N):
        if dp[idx][cnt] == 0:
            continue
        char = S[idx]
        if char == "(":
            dp[idx + 1][cnt + 1] += dp[idx][cnt]
            dp[idx + 1][cnt + 1] %= MOD
        elif char == ")":
            if cnt > 0:
                dp[idx + 1][cnt - 1] += dp[idx][cnt]
                dp[idx + 1][cnt - 1] %= MOD
        else:
            dp[idx + 1][cnt + 1] += dp[idx][cnt]
            dp[idx + 1][cnt + 1] %= MOD
            if cnt > 0:
                dp[idx + 1][cnt - 1] += dp[idx][cnt]
                dp[idx + 1][cnt - 1] %= MOD

print(dp[N][0])

File: atcoder.jp/Rust (rustc 1.70.0)/abc312/abc312_d.rs
use proconio::{fastout, input, marker::Chars};

const MOD: i64 = 998244353;

#[fastout]
fn main() {
    input! {
        s: Chars,
    }
    let n = s.len();

    let mut dp = vec![vec![0; n + 1]; n + 1];
    dp[0][0] = 1;

    for i in 0..n {
        for j in 0..n {
            if dp[i][j] == 0 {
                continue;
            }

            if s[i] == '(' {
                dp[i + 1][j + 1] += dp[i][j];
                dp[i + 1][j + 1] %= MOD;
            } else if s[i] == ')' && j > 0 {
                dp[i + 1][j - 1] += dp[i][j];
                dp[i + 1][j - 1] %= MOD;
            } else if s[i] == '?' {
                dp[i + 1][j + 1] += dp[i][j];
                dp[i + 1][j + 1] %= MOD;
                if j > 0 {
                    dp[i + 1][j - 1] += dp[i][j];
                    dp[i + 1][j - 1] %= MOD;
                }
            }
        }
    }

    println!("{}", dp[n][0]);
}
